opm/simulators/wells/WellHelpers.hpp

```diff
@@ -33,6 +40,108 @@ namespace Opm {
+                pinfo_->communication().allgatherv(send, cstring_size,
+                                                   cstrings.data(), sizes.data(),
+                                                   offsets.data());
+                for(std::size_t i = 0; i < sizes.size(); ++i)
+                {
+                    std::string name(cstrings.data()+offsets[i]);
+                    if (name != pinfo_->name())
+                    {
+                        if (pinfo_->communication().rank() == 0)
+                        {
+                            //only one process per well logs, might not be 0 of MPI_COMM_WORLD, though
+                            std::string msg = std::string("Fatal Error: Not all ranks are computing for the same well")
+                                          + " well should be " + pinfo_->name() + " but is "
+                                + name;
+                            OpmLog::debug(msg);
+                        }
+                        consistentWells = false;
+                        break;
+                    }
+                }
+                pinfo_->communication().barrier();
+                // As not all processes are involved here we need to use MPI_Abort and hope MPI kills them all
+                if (!consistentWells)
+                {
+                    MPI_Abort(MPI_COMM_WORLD, 1);
+                }
+#endif
+                B_->mv(x, y);
+
+                if (this->pinfo_->communication().size() > 1)
+                {
+                    // Only do communication if we must.
+                    // The B matrix is basically a component-wise multiplication
+                    // with a vector followed by a parallel reduction. We do that
+                    // reduction to all ranks computing for the well to save the
+                    // broadcast when applying C^T.
+                    using YField = typename Y::block_type::value_type;
+                    assert(y.size() == 1);
+                    this->pinfo_->communication().allreduce<std::plus<YField>>(y[0].container().data(),
+                                                                               y[0].container().size());
+                }
+            }
+
+            //! y = A x
+            template<class X, class Y>
+            void mmv (const X& x, Y& y) const
+            {
+                if (this->pinfo_->communication().size() == 1)
+                {
+                    // Do the same thing as before. The else branch
+                    // produces different rounding errors and results
+                    // slightly different iteration counts / well curves
+                    B_->mmv(x, y);
+                }
+                else
+                {
+                Y temp(y);
+                mv(x, temp); // includes parallel reduction
+                y -= temp;
+                }
+            }
+        private:
+            const Matrix* B_;
+            const ParallelWellInfo* pinfo_;
```

Comment:
The `pinfo_` name should be changed to `pwinfo_`, I confused it with the ParallelInfo objects we pass around in other parts of the code that are an entirely different class (not suggesting we go for full Hungarian notation though...)

Reply:
Alas more merge conflicts....

Reply:
Again, I do not understand why this would cause merge conflicts.